### PR TITLE
feat: archive extraction for zip&tarballs

### PR DESCRIPTION
**Pull request by danlucian**
Add support for decompressing and unarchiving _.zip_ or _.tar_ (including _gzipped_) files.

- [x] Ready for review
- [x] Follows CONTRIBUTING rules
- [ ] Reviewed by Snyk internal team

### CODE REVIEW

File: lib/extract.ts [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+async function handleExtraction(
+  path: FilePath,
+  temporaryDir: FilePath,
+  childArchiveHandler: ExtractionHandler,
+) {
+  const extractionTarget: FilePath = join(
+    temporaryDir,
+    path.includes(temporaryDir)
+      ? relative(temporaryDir, `${path}${EXTRACTED_DIR_FLAG}`)
```
Comment (dekelund) [inline lib/extract.ts:28]: I would rename `EXTRACTED_DIR_FLAG` to `EXTRACTED_DIR_SUFFIX`

Comment (danlucian) [reply lib/extract.ts:28]: Yep, good suggestion! I'll update it right now.

File: lib/extract.ts [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+      onentry: (entry: any) => {
+        const childAbsolutePath = join(extractionTarget, entry.path);
+
+        if (isArchive(childAbsolutePath)) {
+          childArchiveHandler(childAbsolutePath);
+        }
+      },
+    });
+  }
+
```
Comment (dekelund) [inline lib/extract.ts:48]: I guess we want the first if to be an early exit, or use `else if` to indicate that it's one or the other, not both.

Comment (dekelund) [reply lib/extract.ts:48]: (In this case I would say `else if` or `switch` since we are branching out in the execution tree)

Comment (danlucian) [reply lib/extract.ts]: This should be some form of pattern matching, unfortunately we can not use a `switch`. I updated to a specific else branch for the zips. Thanks for spotting this!

**Review by dekelund — APPROVED** [review]

**Pull request merged**